linkerd/proxy/http/src/normalize_uri.rs

```diff
@@ -55,57 +61,76 @@ impl<N> NewNormalizeUri<N> {
 
 impl<T, N> NewService<T> for NewNormalizeUri<N>
 where
-    T: Param<SocketAddr>,
+    T: Param<DefaultAuthority>,
     N: NewService<T>,
 {
     type Service = NormalizeUri<N::Service>;
 
     fn new_service(&mut self, target: T) -> Self::Service {
-        let target_addr = target.param();
+        let DefaultAuthority(default) = target.param();
         let inner = self.inner.new_service(target);
-        NormalizeUri::new(inner, target_addr)
+        NormalizeUri::new(inner, default)
     }
 }
 
 // === impl NormalizeUri ===
 
 impl<S> NormalizeUri<S> {
-    fn new(inner: S, target_addr: SocketAddr) -> Self {
-        let default = http::uri::Authority::from_str(&target_addr.to_string())
-            .expect("SocketAddr must be a valid Authority");
+    fn new(inner: S, default: Option<Authority>) -> Self {
         Self { inner, default }
     }
 }
 
 impl<S, B> tower::Service<http::Request<B>> for NormalizeUri<S>
 where
     S: tower::Service<http::Request<B>>,
+    S::Error: Into<Error>,
 {
     type Response = S::Response;
-    type Error = S::Error;
-    type Future = S::Future;
+    type Error = Error;
+    type Future = future::Either<
+        future::ErrInto<S::Future, Error>,
+        future::Ready<Result<S::Response, Error>>,
+    >;
 
     #[inline]
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), S::Error>> {
-        self.inner.poll_ready(cx)
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
     }
 
     fn call(&mut self, mut req: http::Request<B>) -> Self::Future {
         if let http::Version::HTTP_10 | http::Version::HTTP_11 = req.version() {
             if req.extensions().get::<h1::WasAbsoluteForm>().is_none()
                 && req.uri().authority().is_none()
             {
-                let authority =
-                    h1::authority_from_host(&req).unwrap_or_else(|| self.default.clone());
+                let authority = match h1::authority_from_host(&req).or_else(|| self.default.clone())
+                {
+                    Some(a) => a,
+                    None => return future::Either::Right(future::err(NoAuthority(()).into())),
+                };
+
                 trace!(%authority, "Normalizing URI");
                 h1::set_authority(req.uri_mut(), authority);
             }
         }
 
-        self.inner.call(req)
+        future::Either::Left(self.inner.call(req).err_into())
     }
 }
 
+// === impl NoAuthority ===
+
+impl std::fmt::Display for NoAuthority {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "failed to normalize URI because no authority could be determined"
```

Comment:
it might be worth including the bad URI in this message? but, that would mean carrying it in the error type